pkg/core/storage/badgerdb_store.go

```diff
@@ -0,0 +1,112 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger/v2"
+)
+
+// BadgerDBOptions configuration for BadgerDB.
+type BadgerDBOptions struct {
+	Dir string `yaml:"BadgerDir"`
+}
+
+// BadgerDBStore is the official storage implementation for storing and retrieving
+// blockchain data.
+type BadgerDBStore struct {
+	db *badger.DB
+}
+
+// NewBadgerDBStore returns a new BadgerDBStore object that will
+// initialize the database found at the given path.
+func NewBadgerDBStore(cfg BadgerDBOptions) (*BadgerDBStore, error) {
+	opts := badger.DefaultOptions(cfg.Dir) // should be exposed via BadgerDBOptions if anything needed
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BadgerDBStore{
+		db: db,
+	}, nil
+}
+
+// Batch implements the Batch interface and returns a badgerdb
+// compatible Batch.
+func (b *BadgerDBStore) Batch() Batch {
+	return newMemoryBatch()
+}
+
+// Delete implements the Store interface.
+func (b *BadgerDBStore) Delete(key []byte) error {
+	return b.db.Update(func(txn *badger.Txn) error {
+		return txn.Delete(key)
+	})
+}
+
+// Get implements the Store interface.
+func (b *BadgerDBStore) Get(key []byte) ([]byte, error) {
+	var val []byte
+	err := b.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err == badger.ErrKeyNotFound {
+			return ErrKeyNotFound
+		}
+		val, err = item.ValueCopy(nil)
+		return err
+	})
+	return val, err
+}
+
+// Put implements the Store interface.
+func (b *BadgerDBStore) Put(key, value []byte) error {
+	return b.db.Update(func(txn *badger.Txn) error {
+		err := txn.Set(key, value)
+		return err
+	})
+}
+
+// PutBatch implements the Store interface.
+func (b *BadgerDBStore) PutBatch(batch Batch) error {
+	wb := b.db.NewWriteBatch()
+	defer wb.Cancel()
+
+	for k, v := range batch.(*MemoryBatch).mem {
+		err := wb.Set([]byte(k), v)
+		if err != nil {
+			return err
+		}
+	}
+	for k := range batch.(*MemoryBatch).del {
+		err := wb.Delete([]byte(k))
+		if err != nil {
+			return err
+		}
+	}
+	return wb.Flush()
+}
+
+// Seek implements the Store interface.
+func (b *BadgerDBStore) Seek(key []byte, f func(k, v []byte)) {
+	err := b.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.DefaultIteratorOptions)
```

Comment:
`IteratorOptions` has `Prefix` field. Doesn't it do what we need?
https://pkg.go.dev/github.com/dgraph-io/badger?tab=doc#IteratorOptions

Reply:
Not quite, so we can't get reed of `it.Seek(key)` here. See https://github.com/dgraph-io/badger#prefix-scans